components/raftstore/src/store/peer.rs

```diff
@@ -5161,6 +5169,21 @@ where
             send_msg.set_end_key(region.get_end_key().to_vec());
         }
 
+        match msg.get_msg_type() {
+            MessageType::MsgAppendResponse | MessageType::MsgHeartbeatResponse => {
+                let mut ext_msg = ExtraMessage::default();
+                ext_msg.set_type(ExtraMessageType::MsgTracePeerAvailabilityInfo);
+                ext_msg.miss_data = self.peer.get_is_witness();
+                ext_msg.safe_ts = if ext_msg.miss_data {
+                    0
+                } else {
+                    self.read_progress.safe_ts()
+                };
+                send_msg.set_extra_msg(ext_msg);
```

Comment:
If the MsgHeartbeatResponse message has extra_msg, https://github.com/tikv/tikv/blob/6dcd0b3d66945583a133e60838533a5d12b3a487/components/raftstore/src/store/fsm/peer.rs#L2320 will return OK and not [execute the `step`](https://github.com/tikv/tikv/blob/6dcd0b3d66945583a133e60838533a5d12b3a487/components/raftstore/src/store/fsm/peer.rs#L2365).

Reply:
Sorry fixed before, but may be lost by one force push :(